experiments/simbricks/runtime/common.py

```diff
@@ -20,17 +20,36 @@
 # TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 # SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-import simbricks.exectools as exectools
+# Allow type annotation of class to be used in its own constructor
+from __future__ import annotations
+
+from abc import abstractmethod
 import shutil
 import pathlib
+import typing as tp
+
+from simbricks.experiment.experiment_environment import ExpEnv
+from simbricks.experiment.experiment_output import ExpOutput
+from simbricks.experiments import Experiment
+import simbricks.exectools as exectools
+
 
 class Run(object):
-    def __init__(self, experiment, index, env, outpath, prereq=None):
+    """Defines a single execution run for an experiment."""
+
+    def __init__(
+        self,
+        experiment: Experiment,
+        index: int,
+        env: ExpEnv,
+        outpath: str,
+        prereq: tp.Optional[Run] = None
+    ):
         self.experiment = experiment
         self.index = index
         self.env = env
         self.outpath = outpath
-        self.output = None
+        self.output: tp.Optional[ExpOutput] = None
```

Comment:
I don't understand what you mean here by 'members before the init function'. Could you explain this further?

Reply:
I was basically wondering if something like this would be  "the right way" and avoid the inline annotation?
```
class Run(object):
    output: tp.Optional[ExpOutput]
    def __init__( ....
```

Reply:
Ah I see looks like that only changed in recent version of python with the introduction of `ClassVar` (https://docs.python.org/3/library/typing.html#typing.ClassVar) then fine to just leave it as it is. :-)

Reply:
I did not know this but since we are targeting Python >= 3.7, I'd much prefer your proposed notation since it removes clutter from `init` and one can easily see all class attributes. If you're ok, I would change this for type annotations added in this PR.

Reply:
Sounds good, if that works I also like that much better. :-)